### PR TITLE
MGMT-9344: Return non-zero exit when subsystem test fails

### PR DESCRIPTION
**Pull request by adriengentil**
As part of moving the subsystem test to prow, we need to know when the
step that runs the test fails.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -61,14 +61,20 @@ endif
...
-	-$(MAKE) _test TEST_SCENARIO=subsystem TIMEOUT=30m TEST="$(or $(TEST),./subsystem/...)"
+
+_post_subsystem:
 	$(DOCKER_COMPOSE) logs dhcpd > dhcpd.log
 	$(DOCKER_COMPOSE) logs wiremock > wiremock.log
 	$(DOCKER_COMPOSE) down
 
+subsystem: build-image
+	$(MAKE) _pre_subsystem
+	$(MAKE) _test TEST_SCENARIO=subsystem TIMEOUT=30m TEST="$(or $(TEST),./subsystem/...)" || ($(MAKE) _post_subsystem && /bin/false)
```
Comment (osherdp) [inline Makefile:75]: it means we only cleanup on success?

File: Makefile [hunk 2]
```diff
@@ -61,14 +61,20 @@ endif
 unit-test:
 	$(MAKE) _test TEST_SCENARIO=unit TIMEOUT=30m TEST="$(or $(TEST),$(shell go list ./... | grep -v subsystem))" || (docker kill postgres && /bin/false)
 
-subsystem: build-image
+_pre_subsystem:
```
Comment (omertuc) [inline Makefile:64]: target should be PHONY

File: Makefile [hunk 3]
```diff
@@ -61,14 +61,20 @@ endif
...
 	-docker image rm subsystem_agent
 	$(DOCKER_COMPOSE) up --build -d dhcpd wiremock
-	-$(MAKE) _test TEST_SCENARIO=subsystem TIMEOUT=30m TEST="$(or $(TEST),./subsystem/...)"
+
+_post_subsystem:
 	$(DOCKER_COMPOSE) logs dhcpd > dhcpd.log
 	$(DOCKER_COMPOSE) logs wiremock > wiremock.log
 	$(DOCKER_COMPOSE) down
 
+subsystem: build-image
```
Comment (osherdp) [inline Makefile:73]: in your release PR this will allow you to use "public" makefile targets (and not ones like ``_<name>``)?

File: Makefile [hunk 4]
```diff
@@ -61,14 +61,20 @@ endif
 unit-test:
 	$(MAKE) _test TEST_SCENARIO=unit TIMEOUT=30m TEST="$(or $(TEST),$(shell go list ./... | grep -v subsystem))" || (docker kill postgres && /bin/false)
 
-subsystem: build-image
+_pre_subsystem:
 	-docker image rm subsystem_agent
 	$(DOCKER_COMPOSE) up --build -d dhcpd wiremock
-	-$(MAKE) _test TEST_SCENARIO=subsystem TIMEOUT=30m TEST="$(or $(TEST),./subsystem/...)"
+
+_post_subsystem:
```
Comment (omertuc) [inline Makefile:68]: same here

File: Makefile (hunk 1 above)
Comment (osherdp) [reply Makefile:75]: not only cleanup, but also gathering

**Review by osherdp — CHANGES_REQUESTED** [review]

File: Makefile [hunk 5]
```diff
@@ -61,14 +61,20 @@ endif
...
 	$(DOCKER_COMPOSE) logs dhcpd > dhcpd.log
 	$(DOCKER_COMPOSE) logs wiremock > wiremock.log
 	$(DOCKER_COMPOSE) down
 
+subsystem: build-image
+	$(MAKE) _pre_subsystem
```
Comment (omertuc) [inline Makefile:74]:
```suggestion
subsystem: build-image _pre_subsystem
```

**Comment by omertuc** [comment]
How is splitting up this target into pre, do, post achieving the goal of the PR?

**Comment by osherdp** [comment]
> How is splitting up this target into pre, do, post achieving the goal of the PR?

I think because it used to be:
```
-$(MAKE) _test TEST_SCENARIO=subsystem TIMEOUT=30m TEST="$(or $(TEST),./subsystem/...)"
```

so the minus sign at start of the line made it to ignore return code

File: Makefile (hunk 1 above)
Comment (adriengentil) [reply Makefile:75]: the trick is in the `|| ($(MAKE) _post_subsystem && /bin/false)`.
If `_test` target fails, the command after the `||` will be executed and since it return a non-zero (`/bin/false`), make will stop the process.

We use the same logic to run the unit tests: https://github.com/openshift/assisted-installer-agent/blob/master/Makefile#L44

@omertuc I think it answers your question

Comment (osherdp) [reply Makefile:75]: sorry, I meant "we only cleanup on failure"
not sure it's that critical though, just an observation

File: Makefile (hunk 3 above)
Comment (adriengentil) [reply Makefile:73]: I'm not sure to understand your question, but I don't change the interface so will we still launch the subsystem tests using `make subsystem`

File: Makefile (hunk 1 above)
Comment (osherdp) [reply Makefile:75]: ah no, all good
please ignore

Comment (omertuc) [reply Makefile:75]: I get what you're doing but I feel it's a bit hard to follow. rather than short-circuiting / `/bin/false` tricks, maybe we can do something like this pseudocode:

```
// Run the tests, record the result
result = run_tests()

// Cleanup, ignore cleanup failures
cleanup || true

// Return the test result
exit $result
```

Comment (adriengentil) [reply Makefile:75]: Each statement in a makefile is executed in a sub shell which makes things a bit awkward, but I just discovered `.ONESHELL` which could makes things simpler.

https://www.gnu.org/software/make/manual/html_node/One-Shell.html

Comment (osherdp) [reply Makefile:75]: make sure to add a comment when using ``.ONESHELL``, because I think it's not a known feature

**Comment by adriengentil** [comment]
@omertuc @osherdp I updated my PR, all the statements are now executed in the same shell (thanks to the backslashes) in the `subsystem` target. I think it's cleaner than before, what do you think?

(I mentioned to use `.ONESCRIPT` in a post above, but it applies to the whole makefile, and as @osherdp pointed out, it can be confusing because it's not a well known feature, so I think it's a good idea to use it)

**Review by osherdp — APPROVED** [review]
yes, it's cleaner
